src/internal_modules/roc_sndio/wav_sink.cpp

```diff
@@ -0,0 +1,268 @@
+    , buffer_(arena)
+    , buffer_size_(0)
+    , valid_(false) {
+    BackendMap::instance();
+
+    if (config.sample_spec.num_channels() == 0) {
+        roc_log(LogError, "wav sink: # of channels is zero");
+        return;
+    }
+
+    if (config.latency != 0) {
+        roc_log(LogError, "wav sink: setting io latency not supported by wav backend");
+        return;
+    }
+
+    frame_length_ = config.frame_length;
+
+    if (frame_length_ == 0) {
+        roc_log(LogError, "wav sink: frame length is zero");
+        return;
+    }
+
+    valid_ = true;
+}
+
+WavSink::~WavSink() {
+    close_();
+}
+
+bool WavSink::is_valid() const {
+    return valid_;
+}
+
+bool WavSink::open(const char* path) {
+    roc_panic_if(!valid_);
+
+    roc_log(LogDebug, "wav sink: opening: path=%s", path);
+
+    if (output_file_ != NULL) {
+        roc_panic("wav sink: can't call open() more than once");
+    }
+
+    if (!open_(path)) {
+        return false;
+    }
+
+    if (!setup_buffer_()) {
+        return false;
+    }
+
+    return true;
+}
+
+DeviceType WavSink::type() const {
+    return DeviceType_Sink;
+}
+
+DeviceState WavSink::state() const {
+    return DeviceState_Active;
+}
+
+void WavSink::pause() {
+    // no-op
+}
+
+bool WavSink::resume() {
+    return true;
+}
+
+bool WavSink::restart() {
+    return true;
+}
+
+audio::SampleSpec WavSink::sample_spec() const {
+    roc_panic_if(!valid_);
+
+    if (!output_file_) {
+        roc_panic("wav sink: sample_spec(): non-open output file or device");
+    }
+
+    if (header_.num_channels() == 1) {
+        return audio::SampleSpec(size_t(header_.sample_rate()),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Mono);
+    }
+
+    if (header_.num_channels() == 2) {
+        return audio::SampleSpec(size_t(header_.sample_rate()),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Stereo);
+    }
+
+    roc_panic("wav sink: unsupported channel count");
+}
+
+core::nanoseconds_t WavSink::latency() const {
+    roc_panic_if(!valid_);
+
+    if (!output_file_) {
+        roc_panic("wav sink: latency(): non-open output file");
+    }
+
+    return 0;
+}
+
+bool WavSink::has_latency() const {
+    roc_panic_if(!valid_);
+
+    if (!output_file_) {
+        roc_panic("wav sink: has_latency(): non-open output file");
+    }
+
+    return false;
+}
+
+bool WavSink::has_clock() const {
+    roc_panic_if(!valid_);
+
+    if (!output_file_) {
+        roc_panic("wav sink: has_clock(): non-open output file");
+    }
+
+    return false;
+}
+
+void WavSink::write(audio::Frame& frame) {
+    roc_panic_if(!valid_);
+
+    const audio::sample_t* frame_data = frame.samples();
+    size_t frame_size = frame.num_samples();
+
+    audio::sample_t* buffer_data = buffer_.data();
+    size_t buffer_pos = 0;
```

Comment:
Extra buffer is not used, but I'm updating the header every time audio data is appended to the file. Otherwise the header would be _wrong_ but not invalid.